packages/headless/src/features/insight-user-actions/insight-user-actions-preprocessing.ts

```diff
@@ -0,0 +1,339 @@
+import {
+  UserAction,
+  UserActionsState,
+  UserActionTimeline,
+  UserActionType,
+  UserSession,
+} from './insight-user-actions-state';
+
+const SECOND = 1000;
+const MINUTE = SECOND * 60;
+const SESSION_INACTIVITY_THRESHOLD_IN_MS = 30 * MINUTE;
+const NUMBER_OF_PRECEDING_SESSIONS = 2;
+const NUMBER_OF_FOLLOWING_SESSIONS = 2;
+
+export interface RawUserAction {
+  name: string;
+  value: string;
+  time: string;
+}
+
+/**
+ * Preprocesses the raw user actions data from the API into a user actions timeline with the following steps:
+ * - The data is first sorted by most recent and mapped into user actions.
+ * - The actions are then split into sessions based on the ticket creation date.
+ * - The sessions are then stored in the timeline object either as preceding, following or the case creation session.
+ * - The timeline is then filtered to exclude custom actions passed in the state `excludedCustomActions` and empty search queries.
+ * Note: The filtering is done at the end to avoid filtering out actions that would lead to incorrectly split a session into two by creating a gap.
+ * @param state {UserActionsState} - The state of the user actions
+ * @param actions {rawUserAction[]} - The raw user actions array to preprocess
+ * @returns {UserActionTimeline}
+ */
+export const preprocessActionsData = (
+  state: UserActionsState,
+  actions: Array<RawUserAction>
+): UserActionTimeline => {
+  if (!state.ticketCreationDate || actions.length === 0) {
+    return {
+      precedingSessions: [],
+      session: undefined,
+      followingSessions: [],
+    };
+  }
+
+  const ticketCreationdate = new Date(state.ticketCreationDate);
+  const ticketCreationTimestamp = ticketCreationdate.getTime();
+  const excludedCustomActions = state.excludedCustomActions;
+
+  const mappedAndSortedActions = mapAndSortActionsByMostRecent(actions);
+
+  const timeline = splitActionsIntoTimelineSessions(
+    mappedAndSortedActions,
+    ticketCreationTimestamp
+  );
+
+  const filteredTimeline = filterTimelineActions(
+    timeline,
+    excludedCustomActions
+  );
+
+  return filteredTimeline;
+};
+
+const mapRawActionToUserAction = (rawAction: RawUserAction): UserAction => {
+  const actionData = JSON.parse(rawAction.value);
+  return {
+    actionType: rawAction.name as UserActionType,
+    timestamp: Number(rawAction.time),
+    eventData: {
+      type: actionData.event_type,
+      value: actionData.event_value,
+    },
+    cause: actionData.cause,
+    searchHub: actionData.origin_level_1,
+    document: {
+      title: actionData.title,
+      uriHash: actionData.uri_hash,
+      // eslint-disable-next-line @cspell/spellchecker
+      contentIdKey: actionData.c_contentidkey,
+      // eslint-disable-next-line @cspell/spellchecker
+      contentIdValue: actionData.c_contentidvalue,
```

Comment:
this needs to be improved, 
the view events have the following structure:
```
  {
    "name": "VIEW",
    "value": "{\"content_id_key\":\"@clickableuri\",\"content_id_value\":\"https://customization-velocity-7501-dev-ed.scratch.live-preview.salesforce-experience.com/s/?app=commeditor&view=editor&formFactor=LARGE&viewid=e85c3fcd-4517-4d71-96e1-a3a706d04065#t=All&sort=relevancy\",\"language\":\"en-US\",\"title\":\"Home\"}",
    "time": "1699478842750"
  },
```

now how the click uri is stored in `content_id_value` and this not reflected in the mapped action  and we need this value to display the view event as a link.

Reply:
fixed that!